Sources/BSON/BSONCode.swift

```diff
@@ -2,6 +2,46 @@ import NIO
 
 /// A struct to represent the BSON Code type.
 public struct BSONCode: Equatable, Hashable {
+    /*
+     * Initializes a `BSONCode` from ExtendedJSON.
+     *
+     * Parameters:
+     *   - `json`: a `JSON` representing the canonical or relaxed form of ExtendedJSON for `Code`.
+     *   - `keyPath`: an array of `String`s containing the enclosing JSON keys of the current json being passed in.
+     *              This is used for error messages.
+     *
+     * Returns:
+     *   - `nil` if the provided value is not a `String`.
+     *
+     * Throws:
+     *   - `DecodingError` if `json` is a partial match or is malformed.
+     */
+    internal init?(fromExtJSON json: JSON, keyPath: [String]) throws {
+        switch json {
+        case let .object(obj):
+            // canonical extended JSON
+            guard let value = obj["$code"] else {
+                return nil
+            }
+            guard obj.count == 1 else {
```

Comment:
this one's actually a little tricky: We also have a `CodeWithScope` type, which has both `$code` and `$scope` keys. If we find that the two keys are exactly `$code` and `$scope`, we should actually just return `nil` here. Otherwise we can throw.

Reply:
I think we want to avoid the implicit dependency on the other initializer here if we can and throw in all error cases. It just so happens that `$code` + `$scope` isn't an error case, so that one we treat differently, but the rest we leave the same.

Reply:
okay sounds good, ill edit